src/cpp/server/load_reporter/load_reporter_async_service_impl.h

```diff
@@ -0,0 +1,195 @@
+  explicit LoadReporterAsyncServiceImpl(
+      std::unique_ptr<ServerCompletionQueue> cq);
+  ~LoadReporterAsyncServiceImpl();
+
+  // Starts the working thread.
+  void StartThread();
+
+  // Not copyable nor movable.
+  LoadReporterAsyncServiceImpl(const LoadReporterAsyncServiceImpl&) = delete;
+  LoadReporterAsyncServiceImpl& operator=(const LoadReporterAsyncServiceImpl&) =
+      delete;
+
+ private:
+  class ReportLoadHandler;
+
+  // A tag that can be called with a bool argument. It's tailored for
+  // ReportLoadHandler's use. Before being used, it should be constructed with a
+  // method of ReportLoadHandler and a shared pointer to the handler. The
+  // shared pointer will be moved to the invoked function and the function can
+  // only be invoked once. That makes ref counting of the handler easier,
+  // because the shared pointer is not bound to the function and can be gone
+  // once the invoked function returns (if not used any more).
+  class CallableTag {
+   public:
+    using HandlerFunction =
+        std::function<void(std::shared_ptr<ReportLoadHandler>, bool)>;
+
+    CallableTag() {}
+
+    CallableTag(HandlerFunction func,
+                std::shared_ptr<ReportLoadHandler> handler)
+        : handler_function_(std::move(func)), handler_(std::move(handler)) {
+      GPR_ASSERT(handler_function_ != nullptr);
+      GPR_ASSERT(handler_ != nullptr);
+    }
+
+    // Runs the tag. This should be called only once. The handler is no longer
+    // owned by this tag after this method is invoked.
+    void Run(bool ok);
+
+    // Releases and returns the shared pointer to the handler.
+    std::shared_ptr<ReportLoadHandler> ReleaseHandler() {
+      return std::move(handler_);
+    }
+
+   private:
+    HandlerFunction handler_function_ = nullptr;
+    std::shared_ptr<ReportLoadHandler> handler_;
+  };
+
+  // Each handler takes care of one load reporting stream. It contains
+  // per-stream data and it will access the members of the parent class (i.e.,
+  // LoadReporterAsyncServiceImpl) for service-wide data (e.g., the load data).
+  class ReportLoadHandler
+      : public std::enable_shared_from_this<ReportLoadHandler> {
+   public:
+    // Instantiates a ReportLoadHandler and requests the next load reporting
+    // call. The handler object will manage its own lifetime, so no action is
+    // needed from the caller any more regarding that object.
+    static void CreateAndStart(ServerCompletionQueue* cq,
+                               LoadReporterAsyncServiceImpl* service,
+                               LoadReporter* load_reporter);
+
+    // This ctor is public because we want to use std::make_shared<> in
+    // CreateAndStart(). This ctor shouldn't be used elsewhere.
+    ReportLoadHandler(ServerCompletionQueue* cq,
+                      LoadReporterAsyncServiceImpl* service,
+                      LoadReporter* load_reporter);
+
+   private:
+    // After the handler has a call request delivered, it starts reading the
```

Comment:
If you have already use enable_shared_from_this<> and shared_from_this(), why do you still need to pass the shared_ptr<> self to all the functions?

Reply:
Good question!

Ultimately we want to simplify the refcounting of the handlers, which basically means the handler callback doesn't need to do anything if it doesn't have a next step to do (e.g., the `OnFinishDone()` callback is the end of a series of operations). But the problem is that the `std::function` will always hold an extra ref if we `std::bind()` the shared pointer to the callback function when creating the tag. That means we have to reset the function to release that extra ref if the callback doesn't have any next step, which doesn't meet our goal. So we add `CallableTag` to temporarily hold the ref for the half-bound closure and feed the ref to that closure when it's invoked. As a result, we avoid binding the shared pointer to any `std::function`. That's why all the handler callbacks have an extra `std::shared_ptr<>` parameter now. 

`shared_from_this()` solves a simpler problem - creating the ref. But now that we have the extra parameter, it can be removed.

Reply:
I don't want to block this PR to check in. Feel free to check in if other reviewers agree your solution is good.

I cannot see the benefit of CallableTag here. IIUC, it is a wrapper of std::bind(Method, shared_ptr). You are stilling doing refcount with shared ptr and manually release the shared ptr when necessary. I cannot see any difference between your current code and to use a std::function directly.

Reply:
After discussion offline, this is resolved.

BTW, I'd like to append some comment after I read "Item 13: Use objects to manage resources" in Effective C++. If we do choose to use `std::shared_ptr` here, moving the refcounting complexity from the "end" (like, manually releasing the ref in the last callback) to the "beginning" (using something like `CallableTag`) is beneficial because it eliminates the effort to make sure the resource is always released, which is hard in case of exceptions being thrown before the resource is released or other situations that cause early returning.